Editor/AGS.Native/agsnative.cpp

```diff
@@ -3827,60 +3827,67 @@ PInteractionScripts convert_interaction_scripts(Interactions ^interactions)
 }
 
-void convert_room_interactions_to_native()
+void convert_room_interactions_to_native(Room ^room, RoomStruct &rs)
 {
-	Room ^roomBeingSaved = TempDataStorage::RoomBeingSaved;
-    thisroom.EventHandlers = convert_interaction_scripts(roomBeingSaved->Interactions);
-	for (int i = 0; i < roomBeingSaved->Hotspots->Count; ++i) 
+    rs.EventHandlers = convert_interaction_scripts(room->Interactions);
+	for (int i = 0; i < room->Hotspots->Count; ++i)
 	{
-        thisroom.Hotspots[i].EventHandlers = convert_interaction_scripts(roomBeingSaved->Hotspots[i]->Interactions);
+        rs.Hotspots[i].EventHandlers = convert_interaction_scripts(room->Hotspots[i]->Interactions);
 	}
-    for (int i = 0; i < roomBeingSaved->Objects->Count; ++i)
+    for (int i = 0; i < room->Objects->Count; ++i)
 	{
-        thisroom.Objects[i].EventHandlers = convert_interaction_scripts(roomBeingSaved->Objects[i]->Interactions);
+        rs.Objects[i].EventHandlers = convert_interaction_scripts(room->Objects[i]->Interactions);
 	}
-    for (int i = 0; i < roomBeingSaved->Regions->Count; ++i)
+    for (int i = 0; i < room->Regions->Count; ++i)
 	{
-        thisroom.Regions[i].EventHandlers = convert_interaction_scripts(roomBeingSaved->Regions[i]->Interactions);
+        rs.Regions[i].EventHandlers = convert_interaction_scripts(room->Regions[i]->Interactions);
 	}
 }
 
 
 
 #pragma unmanaged
 
-
-void save_room_file(const AGSString &path)
-{
-    thisroom.DataVersion = kRoomVersion_Current;
-    copy_room_palette_to_global_palette();
-
-    calculate_walkable_areas();
-
-    thisroom.BackgroundBPP = thisroom.BgFrames[0].Graphic->GetBPP();
+// Fixups and saves the native room struct into the file
+void save_room_file(RoomStruct &rs, const AGSString &path)
+{
+    rs.DataVersion = kRoomVersion_Current;
+
+    // If no backgrounds are present, generate empty ones
+    for (size_t i = 0; i < rs.BgFrameCount; ++i)
+        if (!rs.BgFrames[i].Graphic) // FIXME use of thisgame.color_depth
+            rs.BgFrames[i].Graphic.reset(BitmapHelper::CreateClearBitmap(rs.Width, rs.Height, 0, thisgame.color_depth * 8));
+    // If no masks are present, generate empty ones
+    if (!rs.WalkAreaMask)
+        rs.WalkAreaMask.reset(BitmapHelper::CreateClearBitmap(rs.Width / rs.MaskResolution, rs.Height / rs.MaskResolution, 0, 8));
+    if (!rs.HotspotMask)
+        rs.HotspotMask.reset(BitmapHelper::CreateClearBitmap(rs.Width / rs.MaskResolution, rs.Height / rs.MaskResolution, 0, 8));
+    if (!rs.RegionMask)
+        rs.RegionMask.reset(BitmapHelper::CreateClearBitmap(rs.Width / rs.MaskResolution, rs.Height / rs.MaskResolution, 0, 8));
+    if (!rs.WalkBehindMask)
+        rs.WalkBehindMask.reset(BitmapHelper::CreateClearBitmap(rs.Width, rs.Height, 0, 8));
+
+    calculate_walkable_areas(rs);
+
+    rs.BackgroundBPP = rs.BgFrames[0].Graphic->GetBPP();
+    for (int i = 0; i < 256; ++i)
        thisroom.Palette[i] = thisroom.BgFrames[0].Palette[i];

```

Comment:
hmm, what happened here with this `^M`. is that a DOS linefeed?

Reply:
Recommited.